### PR TITLE
[PyTorch][Core] Fix CUBLAS GGEMM when weight dims are not divisible by 128

### PR DESCRIPTION
**Pull request by vthumbe1503**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by zhongbozhu — APPROVED** [review]
LGTM

**Comment by zhongbozhu** [comment]
For future reference, this fix PR should be applied to NVFP4 recipe as well.

File: transformer_engine/common/gemm/cublaslt_grouped_gemm.cu [hunk 1]
```diff
@@ -846,6 +850,42 @@ __forceinline__ __device__ int64_t compute_grouped_tensor_offset(const TensorSha
...
+  const int64_t scale_tile_m = static_cast<int64_t>(mxfp8_swizzle::GEMM_SWIZZLED_SCALE_TILE_DIM_Y);
+  const int64_t scale_tile_k = static_cast<int64_t>(mxfp8_swizzle::GEMM_SWIZZLED_SCALE_TILE_DIM_X);
+  // Padded byte size of the swizzled MXFP8 scale_inv for a single tensor with data
+  // shape (first, last). Rowwise scales use rows=first, cols=last; columnwise
+  // scales swap the orientation since they are stored in column-major order.
+  const int64_t scale_rows = rowwise ? first : last;
+  const int64_t scale_cols = rowwise ? last : first;
+  const int64_t padded_rows = ((scale_rows + scale_tile_m - 1) / scale_tile_m) * scale_tile_m;
+  const int64_t scale_blocks_cols = (scale_cols + kMxfp8BlockSize - 1) / kMxfp8BlockSize;
+  const int64_t padded_k = ((scale_blocks_cols + scale_tile_k - 1) / scale_tile_k) * scale_tile_k;
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/common/gemm/cublaslt_grouped_gemm.cu]: I agree, wrong variable names make code harder to maintain. If accurate variable names are too clunky, that's hint the code is not organized logically and you need to redesign.

File: transformer_engine/common/gemm/cublaslt_grouped_gemm.cu [hunk 2]
```diff
@@ -330,6 +331,7 @@ struct GroupedOperandSelection {
   NVTEScalingMode scaling_mode = NVTE_DELAYED_TENSOR_SCALING;
   bool with_gemm_swizzled_scales = false;
   bool trans = false;
+  bool scale_rowwise = true;
```
Comment (timmoon10) [inline transformer_engine/common/gemm/cublaslt_grouped_gemm.cu]: A name like this makes me think you can mix row-wise data and column-wise scale. Really, what we want is a way for the kernel to handle differences between row-wise and column-wise formats. Right now it only matters for MXFP8 scale-invs, but I don't see why it couldn't be relevant for other data formats.
```suggestion
  bool rowwise = true;
```

File: tests/pytorch/test_numerics.py [hunk 3]
```diff
@@ -3156,18 +3166,45 @@ def _make_grouped_tensor_quantized_mxfp8(
...
+        first_dims = None
+    else:
+        first_dims = torch.tensor([t.shape[0] for t in tensors], dtype=torch.int64, device=device)
     return tex.group_quantize(grouped_input, quantizer, len(tensors), first_dims)
 
 
+def _per_tensor_quantize_mxfp8(
+    tensors: List[torch.Tensor],
+    *,
+    is_a: bool,
```
Comment (timmoon10) [inline tests/pytorch/test_numerics.py]: Nit: This variable name is a bit overtuned to the specifics of quantized GEMMs, isn't it?

**Review by timmoon10 — COMMENTED** [review]
Overall this looks good. My only serious suggestion is to change the name of `GroupedOperandSelection.scale_rowwise` to `rowwise`, since that actually changes the intent of the kernel.

File: transformer_engine/common/gemm/cublaslt_grouped_gemm.cu (hunk 2 above)
Comment (vthumbe1503) [reply transformer_engine/common/gemm/cublaslt_grouped_gemm.cu]: Agreed

File: transformer_engine/common/gemm/cublaslt_grouped_gemm.cu (hunk 1 above)
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply transformer_engine/common/gemm/cublaslt_grouped_gemm.cu]: Have made it better now. At the end of the day what we want is to use (128,4) as the padding parameters, with 4 being used along the direction of quantization and 128 being used the other dimension. Have defined the directions as x and y. We already use these as the direction in GEMM for the tiles

File: transformer_engine/common/gemm/cublaslt_grouped_gemm.cu (hunk 2 above)
Comment (vthumbe1503) [reply transformer_engine/common/gemm/cublaslt_grouped_gemm.cu]: Done

File: tests/pytorch/test_numerics.py [hunk 4]
```diff
@@ -3208,12 +3245,15 @@ def test_grouped_gemm_grouped_tensor_mxfp8(
 
     transa = layout[0] == "T"
     transb = layout[1] == "T"
-    grouped_A = _make_grouped_tensor_quantized_mxfp8(A, is_a=True, transposed=transa, device="cuda")
+    a_is_weight = layout in ("TN", "NN")
+    grouped_A = _make_grouped_tensor_quantized_mxfp8(
+        A, is_a=True, transposed=transa, device="cuda", is_weight=a_is_weight
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply tests/pytorch/test_numerics.py]: Made it better

**Review by timmoon10 — DISMISSED** [review]

File: tests/pytorch/test_numerics.py [hunk 5]
```diff
@@ -3138,36 +3138,61 @@ def _make_zero_tokens_grouped_tensor(logical_last_dim, is_a):
...
+        first_dims = None
+    else:
+        first_dims = torch.tensor([t.shape[0] for t in tensors], dtype=torch.int64, device=device)
     return tex.group_quantize(grouped_input, quantizer, len(tensors), first_dims)
 
 
+def _per_tensor_quantize_mxfp8(
+    tensors: List[torch.Tensor],
+    *,
+    rowwise: bool,
```
[the remark this answers is not in the document]
Comment (vthumbe1503) [reply tests/pytorch/test_numerics.py:3178]: If inputs are not swizzled already, then TE ensures to do that before GEMM

[the remark this answers is not in the document]
Comment (vthumbe1503) [reply tests/pytorch/test_numerics.py:3178]: So not needed for a functionality test

**Review by timmoon10 — DISMISSED** [review]

**Review by timmoon10 — APPROVED** [review]
